### PR TITLE
[Enhancement] Decorrelate the NetworkTime calculation from the system clocks

### PR DESCRIPTION
**Pull request by liuyehcf**
## Enhancement

Before this pr, here is how we calculate the NetworkTime
* Sender side record a `send_timestamp` right before sending the packet.
* Receiver side record a `receive_timestamp` right after receiveing the packet(excluding the post-receiveing processing time usage), and send back this time to sender.
* Sender side receive the response containing the `receive_timestamp` and use `receive_timestamp - send_timestamp` to get the final NetworkTime

But here's problem, if the system clocks across the machine are inconsistent, then the time computation hits an exception.

After this pr, the NetworkTime calculation changes to 
* Sender side record a `send_timestamp` right before sending the packet.
* Receiver side record a `receive_timestamp` right after receiveing the packet, and record another `response_timestamp` right before sending the response, and get the `receive_process_time` with `response_timestamp - receive_timestamp`, and send this time to sender.
* Sender side record a `receive_response_time` right after receving the response and use `receive_response_time - send_timestamp - receive_process_time` to get the final NetworkTime.

This approach works fine even the system clock are inconsistent.

## Experiment

1fe/3be with machine specification 16c/64g

Pick up one machine to execute the following shell script to introduce time difference

```sh
#!/bin/bash

# Define the maximum drift in seconds
max_drift=60

while true; do
  # Get the current system time in seconds since the Unix epoch
  current_time=$(date +%s)

  # Generate a random drift value within the specified range
  drift=$(( (RANDOM % (2 * max_drift + 1)) - max_drift ))

  # Calculate the modified time by adding the drift
  modified_time=$(( current_time + drift ))

  # Set the system time to the modified time
  # Note: Changing the system time requires administrative privileges
  # and can have broader implications on the machine's functionality.
  # Exercise caution when modifying the system clock.
  # This code assumes a Unix-like environment.
  if [[ $EUID -eq 0 ]]; then
    date -s "@$modified_time" > /dev/null 2>&1
    if [ $? -eq 0 ]; then
      echo "System time changed to $(date)"
    else
      echo "Error: Failed to set the system time."
    fi
  else
    echo "Error: Insufficient privileges to modify the system time."
  fi

  # Wait for a random interval before introducing the next drift
  wait_time=$(shuf -i 1-10 -n 1)
  sleep $wait_time
done
```

Before this pr, the profile like blow, the quer(tpch-q9) execution time is 2s, but the execution time in profile shows 12min:

image

After this pr, the networkTime return back to normal

image

## What type of PR is this:
- [ ] BugFix
- [ ] Feature
- [x] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

## Checklist:
- [ ] I have added test cases for my bug fix or my new feature
- [ ] This pr will affect users' behaviors
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function

## Bugfix cherry-pick branch check:
- [x] I have checked the version labels which the pr will be auto-backported to the target branch
  - [x] 3.0
  - [x] 2.5
  - [x] 2.4
  - [x] 2.3

### CODE REVIEW

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]


:heart_eyes: **pass** : 0 / 0 (0%)

File: be/src/exec/pipeline/exchange/sink_buffer.cpp [hunk 1]
```diff
@@ -355,7 +356,7 @@ Status SinkBuffer::_try_to_send_rpc(const TUniqueId& instance_id, const std::fun
             } else {
                 _try_to_send_rpc(ctx.instance_id, [&]() {
                     _process_send_window(ctx.instance_id, ctx.sequence);
-                    _update_network_time(ctx.instance_id, ctx.send_timestamp, result.receive_timestamp());
+                    _update_network_time(ctx.instance_id, ctx.send_timestamp, result.receive_process_time());
```
Comment (silverbullet233) [inline be/src/exec/pipeline/exchange/sink_buffer.cpp:359]: If the user is upgrading in grayscale, the result received from the old version of BE does not have this field, should we deal with the compatibility issue?

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]


:heart_eyes: **pass** : 0 / 0 (0%)

File: be/src/exec/pipeline/exchange/sink_buffer.cpp [hunk 2]
```diff
@@ -209,10 +209,11 @@ void SinkBuffer::cancel_one_sinker(RuntimeState* const state) {
 }
 
 void SinkBuffer::_update_network_time(const TUniqueId& instance_id, const int64_t send_timestamp,
-                                      const int64_t receive_timestamp) {
-    _last_receive_time = MonotonicNanos();
+                                      const int64_t receive_process_time) {
+    const int64_t receive_response_time = MonotonicNanos();
```
Comment (fzhedu) [inline be/src/exec/pipeline/exchange/sink_buffer.cpp:213]: renamed to `responsed_timestamp`?

File: be/src/exec/pipeline/exchange/sink_buffer.cpp (hunk 1 above)
Comment (fzhedu) [inline be/src/exec/pipeline/exchange/sink_buffer.cpp:359]: move this line up?

Comment (liuyehcf) [reply be/src/exec/pipeline/exchange/sink_buffer.cpp:359]: I think we don't have to, you always get a wrong profile in this situation no matter how you do.

File: be/src/exec/pipeline/exchange/sink_buffer.cpp (hunk 2 above)
Comment (fzhedu) [reply be/src/exec/pipeline/exchange/sink_buffer.cpp:213]: or `get_response_timestamp`? the receive is confusing with exch reveiver.

**Comment by fzhedu** [comment]
do other case's network time have similar problem? like runtime filter? or between FE and BE?

File: be/src/exec/pipeline/exchange/sink_buffer.cpp (hunk 2 above)
Comment (liuyehcf) [reply be/src/exec/pipeline/exchange/sink_buffer.cpp:213]: Ok, I get it.

File: be/src/exec/pipeline/exchange/sink_buffer.cpp (hunk 1 above)
Comment (liuyehcf) [reply be/src/exec/pipeline/exchange/sink_buffer.cpp:358]: done

Comment (liuyehcf) [reply be/src/exec/pipeline/exchange/sink_buffer.cpp]: We have to always keep these two calcuation methods which is a high cost I think. So maybe we can choose to not keep compatible.

Comment (silverbullet233) [reply be/src/exec/pipeline/exchange/sink_buffer.cpp]: > We have to always keep these two calcuation methods if we take compatibility into consideration which is a pretty high cost I think. So maybe we can choose to not keep compatible, which results in a wrong profile.

I think it's not a big problem and can be acceptable, but we need to tell users in release note that the exchange profile may be abnormal during the upgrade…

**Review by silverbullet233 — APPROVED** [review]

**Review by satanson — APPROVED** [review]

File: be/src/service/internal_service.cpp [hunk 3]
```diff
@@ -141,6 +143,27 @@ template <typename T>
...
+                _done->Run();
+            }
+        }
+
+    private:
+        google::protobuf::Closure* _done;
+        PTransmitChunkResult* _response;
+        const int64_t _receive_timestamp = MonotonicNanos();
+    };
+    google::protobuf::Closure* wrapped_done = new WrapClosure(done, response);
```
Comment (fzhedu) [inline be/src/service/internal_service.cpp:165]: leak memory, pls unique_ptr

File: be/src/exec/pipeline/exchange/sink_buffer.cpp (hunk 1 above)
Comment (fzhedu) [reply be/src/exec/pipeline/exchange/sink_buffer.cpp]: would it may cause crash?

**Comment by liuyehcf** [comment]
> do other case's network time have similar problem? like runtime filter? or between FE and BE?

as far as I know, only Exchange has network time?

**Comment by wanpengfei-git** [comment]
@Mergifyio backport branch-2.5

**Comment by wanpengfei-git** [comment]
@Mergifyio backport branch-2.4

**Comment by liuyehcf** [comment]
This issue will be fixed at the following version:
* 2.3.14
* 2.4.6
* 2.5.7
* 3.0.2
* newer version

**Pull request merged**